src/persist-client/src/iter.rs

```diff
@@ -0,0 +1,700 @@
+    /// and smaller keys will be popped before larger keys.
+    next_kvt: Reverse<Option<(&'a [u8], &'a [u8], T)>>,
+    /// Normally the KVT in `next_kvt` is actually the next thing that will be returned by the
+    /// iterator -- unless we've reached the end of a part and there's a later part in the same
+    /// run, in which case we have no idea what's next.
+    can_poll: bool,
+    /// The index of the corresponding iterator.
+    index: usize,
+    /// Whether / not the iterator for the part is the last in its run, or whether there may be
+    /// iterators for the same part in the future.
+    last_in_run: bool,
+    _phantom: PhantomData<D>,
+}
+
+impl<'a, T: Timestamp + Codec64 + Lattice, D: Codec64 + Semigroup> PartRef<'a, T, D> {
+    fn update_peek(&mut self, iter: &mut ConsolidationPartIter<'a, T, D>) {
+        let peek = iter.peek();
+        match peek {
+            Some((k, v, t, _)) => {
+                self.next_kvt = Reverse(Some((k, v, t)));
+                self.can_poll = true;
+            }
+            None if self.last_in_run => {
+                self.next_kvt = Reverse(None);
+                self.can_poll = true;
+            }
+            None => {
+                self.can_poll = false;
+            }
+        }
+    }
+
+    fn pop(&mut self, from: &mut [ConsolidationPartIter<'a, T, D>]) -> Option<TupleRef<'a, T, D>> {
+        let iter = &mut from[self.index];
+        let popped = iter.next();
+        self.update_peek(iter);
+        popped
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ConsolidatingIter<'a, T: Timestamp, D> {
+    parts: Vec<ConsolidationPartIter<'a, T, D>>,
+    heap: BinaryHeap<PartRef<'a, T, D>>,
+    state: Option<TupleRef<'a, T, D>>,
+    drop_stash: &'a mut Option<Tuple<T, D>>,
+    _phantom: PhantomData<fn() -> D>,
+}
+
+impl<'a, T, D> ConsolidatingIter<'a, T, D>
+where
+    T: Timestamp + Codec64 + Lattice,
+    D: Codec64 + Semigroup,
+{
+    pub fn new(
+        init_state: Option<TupleRef<'a, T, D>>,
+        drop_stash: &'a mut Option<Tuple<T, D>>,
+    ) -> Self {
+        Self {
+            parts: vec![],
+            heap: BinaryHeap::new(),
+            state: init_state,
+            drop_stash,
+            _phantom: Default::default(),
+        }
+    }
+
+    fn push(&mut self, mut iter: ConsolidationPartIter<'a, T, D>, last_in_run: bool) {
+        let mut part_ref = PartRef {
+            next_kvt: Reverse(None),
+            can_poll: false,
+            index: self.parts.len(),
+            last_in_run,
+            _phantom: Default::default(),
+        };
+        part_ref.update_peek(&mut iter);
+        self.parts.push(iter);
+        self.heap.push(part_ref);
+    }
+
+    /// Attempt to consolidate as much into the current state as possible.
+    fn consolidate(&mut self) -> Option<TupleRef<'a, T, D>> {
+        loop {
+            let Some(mut part) = self.heap.peek_mut() else {
+                break;
+            };
+            if !part.can_poll {
```

Comment:
there's some subtlety here I want to try saying out loud

this stops the current consolidated iterator because:
1. on a previous loop we first peeked at, and then popped the next tuple from a PartRef
2. we then advanced the PartRef's `next_kvt`, but it contained no more tuples AND the part is not the last part within its run
3. This means there may be the same KVT in the next part from the same run, so the PartRef is given `next_kvt: None, can_poll: false`
4. A `next_kvt` of `None` sorts the `PartRef` to the top of `self.heap` so it is selected on the next iteration
5. `!part.can_poll` is true so we return

Reply:
if all of that checks out, I think I'd prefer returning closer to when we discover that there's no more possible output from this iterator

Reply:
Almost! In step 3, we don't set `next_kvt` to `None`, so we only hit step 5 once all the other iterators that have data for the same kvt have merged it in. (Because of the ordering, we process all the `can_poll: true` entries before the `false` ones.)

This only really makes a difference when two parts end at the same KV: I might exhaust one run, fetch the next part in it, and then immediately exhaust another run without actually making any real progress.

However, that situation is probably not that likely outside of tests, and this logic is already pretty tough to reason about. I'll see if I can get away with simplifying!